### PR TITLE
Parametrize more tests

### PR DESCRIPTION
**Pull request by Yay295**

### CODE REVIEW

**Comment by radarhere** [comment]
My comment is that these changes are not just extracting parameters from tests that start with loops - they're also splitting the parts that aren't in a loop into new tests, and moving outer code into loops. They're also parametrizing not just the inputs, but the expected results as well. I don't think that every loop in a test needs to be converted into parameters.

Happy to hear thoughts from others though.

File: Tests/test_image_filter.py [hunk 1]
```diff
@@ -5,90 +5,110 @@
...
-
-def test_crash():
-
-    # crashes on small images
-    im = Image.new("RGB", (1, 1))
-    im.filter(ImageFilter.SMOOTH)
+        im = hopper(mode)
+        out = im.filter("hello")
+        assert out.mode == im.mode
+        assert out.size == im.size
```
Comment (nulano) [inline Tests/test_image_filter.py:46]: These two asserts are not needed.

File: Tests/test_image_filter.py [hunk 2]
```diff
@@ -5,90 +5,110 @@
...
-    assert rankfilter("1") == (0, 4, 8)
-    assert rankfilter("L") == (0, 4, 8)
-    with pytest.raises(ValueError):
-        rankfilter("P")
-    assert rankfilter("RGB") == ((0, 0, 0), (4, 0, 0), (8, 0, 0))
-    assert rankfilter("I") == (0, 4, 8)
-    assert rankfilter("F") == (0.0, 4.0, 8.0)
+        im.filter(ImageFilter.MinFilter).getpixel((1, 1))
+        im.filter(ImageFilter.MedianFilter).getpixel((1, 1))
+        im.filter(ImageFilter.MaxFilter).getpixel((1, 1))
```
Comment (nulano) [inline Tests/test_image_filter.py:111]: The second and third filter are not tested, I'd parametrize this by the image filter.

**Review by nulano — COMMENTED** [review]
I think these changes look fine, and even found one test that can be improved.

> My comment is that these changes are not just extracting parameters from tests that start with loops - they're also splitting the parts that aren't in a loop into new tests, and moving outer code into loops. [...] I don't think that every loop in a test needs to be converted into parameters.

While I'd agree that not every loop needs to be converted into a parametrized test, these changes look reasonable to me. There are a few tests where the tests now repeatedly load a test input image. Where this is an expensive operation, these can be converted into a [fixture](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session).

> They're also parametrizing not just the inputs, but the expected results as well.

I think this is fine. I've already had a few PRs that add new tests with parametrized expected results merged here, e.g. [test_imagefont:test_getlength](https://github.com/python-pillow/Pillow/blob/cb6c94123e20c76a9e7c36601419ce1415bb69e3/Tests/test_imagefont.py#L167).

File: Tests/test_image_filter.py (hunk 1 above)
Comment (Yay295) [reply Tests/test_image_filter.py]: resolved with 826ab4b17c1c4622a7210b4c72d50606ed2b2d2a

File: Tests/test_image_filter.py (hunk 2 above)
Comment (Yay295) [reply Tests/test_image_filter.py]: resolved with 65694f3fb82bd6b29f1b8750f730ba311e41f8e5

File: Tests/test_image_filter.py [hunk 3]
```diff
@@ -5,90 +5,109 @@
...
-    assert modefilter("L") == (4, 0)
-    assert modefilter("P") == (4, 0)
-    assert modefilter("RGB") == ((4, 0, 0), (0, 0, 0))
-
-
-def test_rankfilter():
-    def rankfilter(mode):
-        im = Image.new(mode, (3, 3), None)
+@pytest.mark.parametrize(
+    "mode,expected",
```
Comment (hugovk) [inline Tests/test_image_filter.py:55]:
```suggestion
    "mode, expected",
```

File: Tests/test_image_filter.py [hunk 4]
```diff
@@ -5,90 +5,109 @@
...
+    #   3 4 5
+    #   6 7 8
+    mod = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
+    im.putdata([0, 0, 1, 2, 5, 1, 5, 2, 0])  # mode=0
+    mod2 = im.filter(ImageFilter.ModeFilter).getpixel((1, 1))
+    assert (mod, mod2) == expected
+
+
+@pytest.mark.parametrize(
+    "mode,expected",
```
Comment (hugovk) [inline Tests/test_image_filter.py:77]:
```suggestion
    "mode, expected",
```

File: Tests/test_image_reduce.py [hunk 5]
```diff
@@ -38,58 +38,64 @@
 gradients_image.load()
 
 
-def test_args_factor():
+@pytest.mark.parametrize(
+    "size,expected",
```
Comment (hugovk) [inline Tests/test_image_reduce.py:42]:
```suggestion
    "size, expected",
```

File: Tests/test_image_reduce.py [hunk 6]
```diff
@@ -38,58 +38,64 @@
...
-    with pytest.raises(ValueError):
-        im.reduce(2, (0, 5, 10, 5))
-    with pytest.raises(ValueError):
-        im.reduce(2, (5, 0, 5, 10))
+    with pytest.raises(error):
+        im.reduce(size)
+
+
+@pytest.mark.parametrize(
+    "size,expected",
```
Comment (hugovk) [inline Tests/test_image_reduce.py:64]:
```suggestion
    "size, expected",
```

File: Tests/test_image_reduce.py [hunk 7]
```diff
@@ -38,58 +38,64 @@
...
+        ((5, 5, 6, 6), (1, 1)),
+    ),
+)
+def test_args_box(size, expected):
+    im = Image.new("L", (10, 10))
+    assert expected == im.reduce(2, size).size
+
+
+@pytest.mark.parametrize(
+    "size,error",
```
Comment (hugovk) [inline Tests/test_image_reduce.py:76]:
```suggestion
    "size, expected_error",
```

File: Tests/test_image_reduce.py [hunk 8]
```diff
@@ -38,58 +38,64 @@
...
+        ((0, 0, 10, 11), ValueError),
+        ((-1, 0, 10, 10), ValueError),
+        ((0, -1, 10, 10), ValueError),
+        ((0, 5, 10, 5), ValueError),
+        ((5, 0, 5, 10), ValueError),
+    ),
+)
+def test_args_box_error(size, error):
+    im = Image.new("L", (10, 10))
+    with pytest.raises(error):
```
Comment (hugovk) [inline Tests/test_image_reduce.py:89]:
```suggestion
def test_args_box_error(size, expected_error):
    im = Image.new("L", (10, 10))
    with pytest.raises(expected_error):
```

File: Tests/test_image_reduce.py [hunk 9]
```diff
@@ -38,58 +38,64 @@
...
-        im.reduce(2.0)
-    with pytest.raises(ValueError):
-        im.reduce((0, 10))
-
-
-def test_args_box():
+@pytest.mark.parametrize(
+    "size,error", ((0, ValueError), (2.0, TypeError), ((0, 10), ValueError))
+)
+def test_args_factor_error(size, error):
```
Comment (hugovk) [inline Tests/test_image_reduce.py:57]:
```suggestion
    "size, expected_error", ((0, ValueError), (2.0, TypeError), ((0, 10), ValueError))
)
def test_args_factor_error(size, expected_error):
```

File: Tests/test_image_reduce.py [hunk 10]
```diff
@@ -38,58 +38,64 @@
...
-        im.reduce(2, (0, 0, 10, 11))
-    with pytest.raises(ValueError):
-        im.reduce(2, (-1, 0, 10, 10))
-    with pytest.raises(ValueError):
-        im.reduce(2, (0, -1, 10, 10))
-    with pytest.raises(ValueError):
-        im.reduce(2, (0, 5, 10, 5))
-    with pytest.raises(ValueError):
-        im.reduce(2, (5, 0, 5, 10))
+    with pytest.raises(error):
```
Comment (hugovk) [inline Tests/test_image_reduce.py:59]:
```suggestion
    with pytest.raises(expected_error):
```

File: Tests/test_image_transform.py [hunk 11]
```diff
@@ -75,23 +75,25 @@ def test_quad(self):
...
-                im.size,
-                Image.Transform.EXTENT,
-                (0, 0, w * 2, h * 2),
-                Image.Resampling.BILINEAR,
-                fillcolor="red",
-            )
-
-            assert transformed.getpixel((w - 1, h - 1)) == pixel
+    @pytest.mark.parametrize(
+        "mode,pixel",
```
Comment (hugovk) [inline Tests/test_image_transform.py:79]:
```suggestion
        "mode, expected_pixel",
```

File: Tests/test_image_transform.py [hunk 12]
```diff
@@ -75,23 +75,25 @@ def test_quad(self):
...
+        (
+            ("RGB", (255, 0, 0)),
+            ("RGBA", (255, 0, 0, 255)),
+            ("LA", (76, 0)),
+        ),
+    )
+    def test_fill(self, mode, pixel):
```
Comment (hugovk) [inline Tests/test_image_transform.py:86]:
```suggestion
    def test_fill(self, mode, expected_pixel):
```

File: Tests/test_image_transform.py [hunk 13]
```diff
@@ -75,23 +75,25 @@ def test_quad(self):
...
+        im = hopper(mode)
+        (w, h) = im.size
+        transformed = im.transform(
+            im.size,
+            Image.Transform.EXTENT,
+            (0, 0, w * 2, h * 2),
+            Image.Resampling.BILINEAR,
+            fillcolor="red",
+        )
+        assert transformed.getpixel((w - 1, h - 1)) == pixel
```
Comment (hugovk) [inline Tests/test_image_transform.py:96]:
```suggestion
        assert transformed.getpixel((w - 1, h - 1)) == expected_pixel
```

File: Tests/test_image_transform.py [hunk 14]
```diff
@@ -239,7 +239,16 @@ def _test_image(self):
         im = hopper("RGB")
         return im.crop((10, 20, im.width - 10, im.height - 20))
 
-    def _test_rotate(self, deg, transpose):
+    @pytest.mark.parametrize(
+        "deg,transpose",
```
Comment (hugovk) [inline Tests/test_image_transform.py:243]:
```suggestion
        "deg, transpose",
```

File: Tests/test_image_transform.py [hunk 15]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
-
-    def test_rotate_180_deg(self):
-        self._test_rotate(180, Image.Transpose.ROTATE_180)
-
-    def test_rotate_270_deg(self):
-        self._test_rotate(270, Image.Transpose.ROTATE_270)
-
-    def _test_resize(self, scale, epsilonscale):
+    @pytest.mark.parametrize(
+        "scale,epsilonscale",
```
Comment (hugovk) [inline Tests/test_image_transform.py:284]:
```suggestion
        "scale, epsilon_scale",
```

File: Tests/test_image_transform.py [hunk 16]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
+    )
+    @pytest.mark.parametrize(
+        "resample,epsilon",
+        (
+            (Image.Resampling.NEAREST, 0),
+            (Image.Resampling.BILINEAR, 2),
+            (Image.Resampling.BICUBIC, 1),
+        ),
+    )
+    def test_resize(self, scale, epsilonscale, resample, epsilon):
```
Comment (hugovk) [inline Tests/test_image_transform.py:301]:
```suggestion
    def test_resize(self, scale, epsilon_scale, resample, epsilon):
```

File: Tests/test_image_transform.py [hunk 17]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
         size_up = int(round(im.width * scale)), int(round(im.height * scale))
         matrix_up = [1 / scale, 0, 0, 0, 1 / scale, 0, 0, 0]
         matrix_down = [scale, 0, 0, 0, scale, 0, 0, 0]
 
-        for resample, epsilon in [
+        transformed = im.transform(size_up, self.transform, matrix_up, resample)
+        transformed = transformed.transform(
+            im.size, self.transform, matrix_down, resample
+        )
+        assert_image_similar(transformed, im, epsilon * epsilonscale)
```
Comment (hugovk) [inline Tests/test_image_transform.py:312]:
```suggestion
        assert_image_similar(transformed, im, epsilon * epsilon_scale)
```

File: Tests/test_image_transform.py [hunk 18]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
+
+    @pytest.mark.parametrize(
+        "x,y,epsilonscale",
```
Comment (hugovk) [inline Tests/test_image_transform.py:315]:
```suggestion
        "x, y, epsilon_scale",
```

File: Tests/test_image_transform.py [hunk 19]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
+        (
+            (0.1, 0, 3.7),
+            (0.6, 0, 9.1),
+            (50, 50, 0),
+        ),
+    )
+    @pytest.mark.parametrize(
+        "resample,epsilon",
```
Comment (hugovk) [inline Tests/test_image_transform.py:323]:
```suggestion
        "resample, epsilon",
```

File: Tests/test_image_transform.py [hunk 20]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
-    def test_resize_2_3x(self):
-        self._test_resize(2.3, 3.7)
-
-    def test_resize_2_5x(self):
-        self._test_resize(2.5, 3.7)
-
-    def _test_translate(self, x, y, epsilonscale):
+        ),
+    )
+    def test_translate(self, x, y, epsilonscale, resample, epsilon):
```
Comment (hugovk) [inline Tests/test_image_transform.py:330]:
```suggestion
    def test_translate(self, x, y, epsilon_scale, resample, epsilon):
```

File: Tests/test_image_transform.py [hunk 21]
```diff
@@ -271,77 +280,65 @@ def _test_rotate(self, deg, transpose):
...
-    def test_translate_0_6(self):
-        self._test_translate(0.6, 0, 9.1)
-
-    def test_translate_50(self):
-        self._test_translate(50, 50, 0)
+        transformed = im.transform(size_up, self.transform, matrix_up, resample)
+        transformed = transformed.transform(
+            im.size, self.transform, matrix_down, resample
+        )
+        assert_image_similar(transformed, im, epsilon * epsilonscale)
```
Comment (hugovk) [inline Tests/test_image_transform.py:341]:
```suggestion
        assert_image_similar(transformed, im, epsilon * epsilon_scale)
```

**Review by hugovk — CHANGES_REQUESTED** [review]
A few little whitespace/naming nits

**Review by hugovk — APPROVED** [review]

**Pull request merged**